var/spack/repos/builtin/packages/openmpi/package.py

```diff
@@ -478,7 +478,11 @@ def delete_mpirun_mpiexec(self):
         # only sensible choice (orterun is still present, but normal
         # users don't know about that).
         if '@1.6: ~legacylaunchers schedulers=slurm' in self.spec:
-            os.remove(self.prefix.bin.mpirun)
-            os.remove(self.prefix.bin.mpiexec)
-            os.remove(self.prefix.bin.shmemrun)
-            os.remove(self.prefix.bin.oshrun)
+            try:
+                # OpenMPI 4.0 does not have shmemrun and oshrun
+                os.remove(self.prefix.bin.mpirun)
+                os.remove(self.prefix.bin.mpiexec)
+                os.remove(self.prefix.bin.shmemrun)
+                os.remove(self.prefix.bin.oshrun)
+            except OSError:
```

Comment:
Not sure whether `try`-`except` or simply `if` is preferred.

Reply:
@JiaweiZhuang the try/except logic is fine but ideally this would be in a loop such that a failure in one script wouldn't result in ignoring the rest:

```
for exe in [self.prefix.bin.mpirun, self.prefix.bin.mpiexec...]:
  try:
    os.remove(exe)
  except OSError: # IMO the error code should be checked here to avoid other possible OSError instances, we likely only want to ignore missing file issues
    tty.debug("File not present: " + exe) # this would be useful if other failures cause a file to be unavailable for unexpected reasons
```

Granted that you organized the files such that this failure won't matter for the case you mentioned, but I think it would make things more robust for the future (along with the other suggestions I made in the code example).

Reply:
Go point. Do you mean `llnl.util.tty.debug()`? Haven't seen it being used in other package's `package.py`, which makes me curious if I should use it.

Reply:
> Do you mean llnl.util.tty.debug()

Yes, in this case I think `debug` is appropriate vs. `warn` because generally it is not a problem when it skips these files (but if something goes wrong it may be worthwhile knowing what was missing)

Reply:
fixed as suggested.